Documentation for transient storage (#15409)
ethereum/solidity: docs/introduction-to-smart-contracts.rst

Since EVM executors can choose to include a transaction or not,
transaction senders cannot abuse the system by setting a low gas price.

.. index:: ! storage, ! memory, ! stack, ! transient storage

.. _locations:

Storage, Transient Storage, Memory and the Stack
================================================

> The Ethereum Virtual Machine has different areas where it can store data with the more

Reviewer: ```suggestion
The Ethereum Virtual Machine has different areas where it can store data with the most
```

> proeminent being storage, transient storage, memory and the stack.

Reviewer 2: ```suggestion
prominent being storage, transient storage, memory and the stack.
```

Store data like derived calculations, caching, and aggregates outside of the contract.
A contract can neither read nor write to any storage apart from its own.

> Similar to storage, there is another data area called **transient storage**, where the main difference

Reviewer 3: Should we add "Transient Storage" to the title of the section?

Also below it talks about "memory" as "the second data area" - if we talked about storage and transient storage already, might make sense to refer to it as the third.

Reviewer 3: Also: weird that this section doesn't mention calldata or returndata or code or anything... (especially ``calldata``, since it's properly exposed on the language level)

Author: I added "Transient Storage" to the title.
Regarding the lack of mention to calldata/returndata/code, I will add a small paragraph about them here...I guess should be simple enough

Reviewer 3: We can also postpone it, it's not really related to this PR itself

Reviewer 3: But yeah, a small paragraph would definitely be enough - they're all read-only, so it'd be enough to mention that there's also a those read-only data locations for passing around arguments and return values across call frames, and accessing code (maybe mentioning that immutables live in code) or so.

Author: I just added a little section about them.

> Similar to storage, there is another data area called **transient storage**, which main difference
> is the fact that it is reset at the end of each transaction. The values stored in this data location
> are persisted across function calls originating from the first call of the transaction. When the transaction
> ends, the transient storage is reset and all values are lost to subsequent transaction calls. Albeit that,
> the cost of reading and writing to transient storage is significantly cheaper than storage.

Reviewer 2: ```suggestion
Similar to storage, there is another data area called **transient storage**, where the main difference
is the fact that it is reset at the end of each transaction.
The values stored in this data location persist only across function calls originating from the first call of the transaction.
When the transaction ends, the transient storage is reset and the values stored there become unavailable to calls in subsequent transactions.
Albeit that, the cost of reading and writing to transient storage is significantly lower than for storage.
```

Reviewer 2: I think that we should include some warning about the consequences of not resetting transient storage. Essentially what we warned about in the blog post. Not sure if this is the best place - it's just an intro after all - but it should be somewhere. Perhaps we should keep it short here but have something more detailed elsewhere where we properly discuss data locations?

We could also use an example - perhaps the reentrancy lock should be one of the examples for beginners?

Author: I will add a brief description of the problem and reference another (to be created) section where transient storage is explored further.
I think the example is a good idea, but should we add it here or in another section?

Author: I added a section briefly describing transient storage and showing a simple reentrancy lock example.

Reviewer 2: > I think the example is a good idea, but should we add it here or in another section?

I was thinking about putting one in the "Solidity by Example" section. But maybe having it in the section about transient storage is enough.

Store data like derived calculations, caching, and aggregates outside of the contract.
A contract can neither read nor write to any storage apart from its own.

Similar to storage, there is another data area called **transient storage**,

> where the main difference is the fact that it is reset at the end of each transaction.

Reviewer: ```suggestion
where the main difference is that it is reset at the end of each transaction.
```

The values stored in this data location persist only across function calls originating
from the first call of the transaction.
When the transaction ends, the transient storage is reset and the values stored there
become unavailable to calls in subsequent transactions.

> Albeit that, the cost of reading and writing to transient storage is significantly lower than for storage.

Reviewer: ```suggestion
Despite this, the cost of reading and writing to transient storage is significantly lower than for storage.
```

but it is not possible to just access arbitrary elements deeper in the stack
without first removing the top of the stack.

Calldata, Returndata and Code
=============================

There are also other data areas which are not as apparent as those discussed previously.
However, they are routinely used during the execution of smart contract transactions.

> The calldata region is the data sent to a transaction as part of a smart contract transaction.

Reviewer 2: Parameters of external functions are also passed through calldata, even within the same transaction.

For example, when creating a contract, calldata would be the constructor code of the new contract.
The Solidity language has the keyword ``calldata`` which can be used to declare

> function input parameters (other than that, they can also be ``memory``).

Reviewer 2: ```suggestion
The parameters of external functions are always initially stored in calldata in an ABI-encoded form and only then decoded into the location specified in their declaration.
If declared as ``memory``, the compiler will eagerly decode them into memory at the beginning of the function,
while marking them as ``calldata`` means that this will be done lazily, only when accessed.
Value types and ``storage`` pointers are decoded directly onto the stack.
```

The returndata is the way a smart contract can return a value after a call.

> In general, Solidity functions use the ``return`` keyword to write values in the returndata area.

Reviewer 2: ```suggestion
In general, external Solidity functions use the ``return`` keyword to ABI-encode values into the returndata area.
```
